SSEを利用して在室者一覧を自動更新するようにした

equal-l2 opened the pull request:
とりあえず普通のJSで書く時と同じ方法でSSEのハンドリングを実装して、loggingストアのactionsに `setUpSSE` として放り込みました。
最初に `setUpSSE` を呼んでやるとイベントハンドラにストアの処理を入れてくれて、あとは裏で勝手に更新してくれます。

※イベントのデータ構造を https://github.com/su-its/ams-backend-nodejs/pull/54 に合わせてあるので、テスト時はバックエンドをそっちからチェックアウトしてください。

TODO:
- この書き方でいいのか（Nuxtにもっといい書き方がないか）
- ~ストアで `baseUrl` を取ってくる方法が知りたい (axiosが使えないので別途必要になる)~

---

closes #24

equal-l2 committed:
SSEを利用して在室者一覧を自動更新するようにした

バックエンドの在室者一覧に変更があったとき、SSEで新しい在室者一覧が届く

ghost commented:
@equal-l2 storeでaxiosがbaseURLじゃない場合、フルパスかけばおｋです

ただ、あまり良い設計ではないので、そこは認識してもらえると良いかと思います。

Stroheim001 commented on store/logging.js:
```diff
@@ -88,5 +88,22 @@ export const actions = {
+        const json = JSON.parse(e.data)
+        commit('SET_IN_ROOM_USERS', json.data)
+      } catch (error) {
+        // console.log('error in parse mesg:', error)
```
@equal-l2 
エラーを握りつぶすのはあまり良い設計とは思えないです。

これが一時的な設計な場合じゃなければ、一応エラーハンドリングしておいてほしいです。

出来ればログとして出しておくと良いかもしれません

equal-l2 replied:
nuxtだとロギングはどうすればいいんでしょう
`console.log`の類はESLintに怒られるので……

equal-l2 replied:
ちなみに、ここで起きるエラーは基本的にバックエンドが送ってくるものが悪いので、エラーハンドリングとしてフロントエンドでできることはせいぜいロギングくらいです

Stroheim001 replied:
この実装が一時的なものでないなら、

ロギングしてないと、エラーが分からなくて困ると思いますが、そうでもないですか？

equal-l2 replied:
ですから、そのロギングの方法をどうすればいいかお聞きしたいんです。
console.logに吐くのでいいんであればESLintを黙らせてそうすればいいだけですが、Nuxtにもっとファンシーなやり方があるのならそちらに従いたいと思いまして。

Stroheim001 replied:
@equal-l2 あ、すみません。下の方のコメントしか読んでませんでしたorz

おそらくなのですが、デバッグ用のロギングと、本番のロギング(実際のnuxtはpm2で動作するのが一般的なので)

で違うので、今はdev用のロギングを考えたいですね。

https://t-salad.com/log4js/

ちなみにこういうの見つけたのですが使えそうですか？

equal-l2 replied:
よく考えたらそもそもvuexとかってブラウザ側で動いてるわけで、ブラウザのconsoleにエラーを出すとかダイアログを出すとか以上のロギングってあんまり意味がなさそうですね。

せっかく調べていただいたのにすみませんが、ここは`console.log`や`console.error`で行くことにします。
(それに伴ってlogging.jsだけ`console.log`禁止を解除するようにしておきます)

equal-l2 replied:
そういえばaxios周りはPromiseがrejectされた場合のハンドリングが何も書いてないようですが、これはどっかでうまくやってくれる感じですか？

equal-l2 replied:
`console.debug`で出力することとしました。

equal-l2 committed:
setUpSSE: baseUrlをnuxt.config.jsから取得するようにした

dotenvモジュールがprocess.envに設定値を置いてくれるのでそれを読めばいい
https://github.com/nuxt-community/dotenv-module#usage

Stroheim001 commented on pages/index.vue:
```diff
@@ -33,6 +33,9 @@ export default {
       room_user: 'logging/inRoomUsers'
     })
+  },
+  created () {
```
そういえばなんですが、別にcreatedでも出来なくはないですがfetchにしておけば、
thisを省略できることと、nuxtのライフサイクル的にデータのfetchは

- fetch
- asyncData

になると思いますが、何故そうしなかったのか意図があったら教えてください。

equal-l2 replied:
正直ライフサイクルのあたりをあんまり勉強してなくて、「createdなら他の準備が終わってないせいでこけることはないだろう」ということで次善策としてcreatedの段階でやってます。

`setUpSSE`自体はSSEのコネクションを開くだけで、データの取得自体はイベントベースで好き勝手に動くので、どこに書いても一緒といえば一緒ですが。

equal-l2 replied:
どこでもいいなら`fetch`で`setUpSSE`をディスパッチしたほうが、新しく`created`を書かなくていいのでわかりやすいかもですね。
そのように修正してもいいですか？

Stroheim001 replied:
@equal-l2 たぶんなんですけど、作法的には、fetchとかのほうが良いと思うし、dataがstoreに入ったことが明示されるかなぁ、と思います。

ベストプラクティスかはわかりませんが

equal-l2 replied:
そのようにしました。

equal-l2 committed:
setUpSSE: エラーが起きた際にdebug出力

equal-l2 committed:
index.vue: setUpSSEをfetch段階で呼び出す

Stroheim001 reviewed (approved):
指摘取り込みお疲れ様でした。

頓珍漢なことを言ってしまったかもしれませんが、その点はすみませんでした。

コードはキレイなので大丈夫だと思います。

h-takeyeah reviewed (commented):
動きました．大丈夫だと思います．レビュー遅くなってすみませんでした．
